libgadget/utils/system.c

```diff
@@ -636,3 +636,93 @@ void gadget_setup_thread_arrays(int * dest, int * srcs[], size_t sizes[], size_t
         sizes[i] = 0;
     }
 }
+
+#ifdef DEBUG
+
+static void
+check_reduce(const void *inputcpy, const void * recvbuf, const int count, MPI_Datatype datatype, MPI_Op op, const int line, const char * file)
+{
+    int i;
+    /* Check that the max/min we got back is larger than or equal to the input*/
+    for(i=0; i < count; i++) {
+        if(op == MPI_MAX) {
+            if(datatype == MPI_INT) {
+                if(((int *) inputcpy)[i] > ((int *) recvbuf)[i])
+                    endrun(12, "MPI_Allreduce with MPI_INT MPI_MAX | MPI_SUM has int %d bad: (input %d > out %d) at %s:%d\n", i, *((int*) inputcpy), *((int*) recvbuf), file, line);
+            }
+            else if (datatype == MPI_LONG) {
+                if(((long *) inputcpy)[i] > ((long *) recvbuf)[i])
+                    endrun(12, "MPI_Allreduce with MPI_LONG MPI_MAX | MPI_SUM has int %d bad: (input %ld > out %ld) at %s:%d\n", i, *((long*) inputcpy), *((long*) recvbuf), file, line);
+            }
+            else if(datatype == MPI_DOUBLE) {
+                if(((double *) inputcpy)[i] > ((double *) recvbuf)[i])
+                    endrun(12, "MPI_Allreduce with MPI_DOUBLE MPI_MAX | MPI_SUM has int %d bad: (input %g > out %g) at %s:%d\n", i, *((double*) inputcpy), *((double*) recvbuf), file, line);
+            }
+        } else if (op == MPI_MIN) {
+            if(datatype == MPI_INT) {
+                if(((int *) inputcpy)[i] < ((int *) recvbuf)[i])
+                    endrun(12, "MPI_Allreduce with MPI_INT MPI_MAX | MPI_SUM has int %d bad: (input %d < out %d) at %s:%d\n", i, *((int*) inputcpy), *((int*) recvbuf), file, line);
+            }
+            else if (datatype == MPI_LONG) {
+                if(((long *) inputcpy)[i] < ((long *) recvbuf)[i])
+                    endrun(12, "MPI_Allreduce with MPI_LONG MPI_MAX | MPI_SUM has int %d bad: (input %ld < out %ld) at %s:%d\n", i, *((long*) inputcpy), *((long*) recvbuf), file, line);
+            }
+            else if(datatype == MPI_DOUBLE) {
+                if(((double *) inputcpy)[i] < ((double *) recvbuf)[i])
+                    endrun(12, "MPI_Allreduce with MPI_DOUBLE MPI_MAX | MPI_SUM has int %d bad: (input %g < out %g) at %s:%d\n", i, *((double*) inputcpy), *((double*) recvbuf), file, line);
+            }
+        }
+    }
+}
+
+int
+MPI_Allreduce_Checked(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype, MPI_Op op, MPI_Comm comm, const int line, const char * file)
```

Comment:
What about CHECKED_MPI_Allreduce?

Reply:
If you feel strongly about it, sure, but I prefer MPI_Allreduce_Checked!